libraries/standard/mqtt/src/iot_mqtt_subscription.c

```diff
@@ -88,108 +142,123 @@ static bool _packetMatch( const IotLink_t * pSubscriptionLink,
 
 /*-----------------------------------------------------------*/
 
-static bool _topicMatch( const IotLink_t * pSubscriptionLink,
-                         void * pMatch )
+static bool _matchTerminalWildcards( const char * pTopicFilter,
+                                     uint16_t nameIndex,
+                                     uint16_t filterIndex,
+                                     uint16_t topicNameLength,
+                                     uint16_t topicFilterLength )
 {
-    bool status = false;
-    uint16_t nameIndex = 0, filterIndex = 0;
+    bool status = false, terminalChar = false;
 
-    /* Because this function is called from a container function, the given link
-     * must never be NULL. */
-    IotMqtt_Assert( pSubscriptionLink != NULL );
+    /* Determine if the terminal character is reached for both topic name and topic
+     * filter for the '#' wildcard. */
+    terminalChar = ( nameIndex == topicNameLength - 1 ) && ( filterIndex == topicFilterLength - 3 );
```

Comment:
Why "- 1" and "- 3"?

Also, what if topicFilterLength is less than three?

Reply:
`topicNameLength - 1` refers to the end of the topic name.  This is the index of the last character.

`topicFilterLength - 3` refers to the character before the wildcard and separator, so before "/#" (see line 160).  The index of this character is 3 before the end of the topic filter.

If `topicFilterLength` is less than 3, then `topicFilterLength - 3` will underflow to about 16535 (`topicFilterLength` is a u16).  This will cause the comparison `( filterIndex == topicFilterLength - 3 )` to evaluate to `false` (`filterIndex` is always in range `[0,topicFilterLength-1]`) and cause this check to be skipped.  That is the correct behavior as this is checking:
> “sport/#” also matches the singular “sport”, since # includes the parent level.

Which requires at least 3 characters ('/#' and another before the separator)